File: soul-planner/soul_planner/scheduler.py
"""Daily planner parser and task scheduler for soul-planner.

Reads ~/soul/docs/daily-planner.md, finds today's date section,
extracts uncompleted tasks from specified blocks, and queues them
into the soul-planner backlog.
"""

from __future__ import annotations

import re
from datetime import date
from pathlib import Path

import aiofiles
import structlog

from soul_planner.db import TaskDB
from soul_planner.models import Priority, TaskCreate

logger = structlog.get_logger("soul-planner.scheduler")

# Map block numbers to priorities
BLOCK_PRIORITY = {
    1: Priority.HIGH,     # BUILD
    2: Priority.NORMAL,   # EXPLORE
    3: Priority.NORMAL,   # SOCIAL
    4: Priority.LOW,      # SCOUT
}

# Month name mapping
MONTHS = {
    "jan": 1, "feb": 2, "mar": 3, "apr": 4, "may": 5, "jun": 6,
    "jul": 7, "aug": 8, "sep": 9, "oct": 10, "nov": 11, "dec": 12,
}

DEFAULT_PLANNER_PATH = Path.home() / "soul" / "docs" / "daily-planner.md"


def _parse_date_from_header(header: str) -> date | None:
    """Extract a date from a day header like 'Day 2 -- Sun Feb 23'."""
    m = re.search(r"(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)\s+(\w+)\s+(\d+)", header)
    if not m:
        return None
    month_str = m.group(1).lower()[:3]
    day_num = int(m.group(2))
    month_num = MONTHS.get(month_str)
    if not month_num:
        return None
    # Assume current year
    return date(date.today().year, month_num, day_num)


def _extract_block_number(header: str) -> int | None:
    """Extract block number from a block header like '### Block 1: BUILD ...'."""
    m = re.match(r"###\s+Block\s+(\d+)", header)
    return int(m.group(1)) if m else None


def parse_planner(content: str, target_date: date, block_filter: int | None = None) -> list[dict]:
    """Parse daily planner content and extract uncompleted tasks.

    Args:
        content: Full text of daily-planner.md
        target_date: Date to find tasks for
        block_filter: If set, only extract tasks from this block number (1-4)

    Returns:
        List of dicts with keys: title, block, priority
    """
    lines = content.split("\n")
    tasks: list[dict] = []

    in_target_day = False
    current_block: int | None = None

    for line in lines:
        # Detect day headers: ## Day N -- Weekday Month Date
        if line.startswith("## Day "):
            header_date = _parse_date_from_header(line)
            in_target_day = header_date == target_date
            current_block = None
            continue

        if not in_target_day:
            continue

        # Detect next day section (exit)
        if line.startswith("---"):
            break

        # Detect block headers
        if line.startswith("### Block "):
            current_block = _extract_block_number(line)
            continue

        # Skip non-block sections (like Evening Review)
        if line.startswith("### ") and not line.startswith("### Block"):
            current_block = None
            continue

        # Only process if we're in a block
        if current_block is None:
            continue

        # Apply block filter
        if block_filter is not None and current_block != block_filter:
            continue

        # Extract uncompleted tasks: - [ ] task text
        m = re.match(r"^- \[ \] (.+)$", line.strip())
        if m:
            title = m.group(1).strip()
            tasks.append({
                "title": title,
                "block": current_block,
                "priority": BLOCK_PRIORITY.get(current_block, Priority.NORMAL),
            })

    return tasks


async def schedule_tasks(
    db: TaskDB,
    planner_path: Path | None = None,
    target_date: date | None = None,
    block_filter: int | None = None,
) -> list[int]:
    """Parse daily planner and queue uncompleted tasks.

    Args:
        db: TaskDB instance
        planner_path: Path to daily-planner.md (default: ~/soul/docs/daily-planner.md)
        target_date: Date to schedule for (default: today)
        block_filter: Only schedule tasks from this block (1-4)

    Returns:
        List of created task IDs
    """
    path = planner_path or DEFAULT_PLANNER_PATH
    today = target_date or date.today()

    # Validate path exists before reading
    if not path.exists():
        logger.error("Planner file not found", path=str(path))
        return []

    async with aiofiles.open(path, encoding="utf-8") as f:
        content = await f.read()
    parsed = parse_planner(content, today, block_filter)

    if not parsed:
        logger.info("No uncompleted tasks found", date=str(today), block=block_filter)
        return []

    # Check for duplicates
    existing = await db.list()
    existing_titles = {t.title for t in existing}

    created_ids: list[int] = []
    for item in parsed:
        if item["title"] in existing_titles:
            logger.debug("Skipping duplicate task", title=item["title"])
            continue

        task = await db.add(TaskCreate(
            title=item["title"],
            description=f"From daily planner, Block {item['block']}",
            priority=item["priority"],
        ))
        created_ids.append(task.id)
        existing_titles.add(item["title"])

    logger.info("Tasks scheduled", count=len(created_ids), date=str(today), block=block_filter)
    return created_ids

File: soul-planner/src/scheduler.rs
/// Daily planner parser and task scheduler for soul-planner.
///
/// Reads ~/soul/docs/daily-planner.md, finds today's date section,
/// extracts uncompleted tasks from specified blocks, and queues them
/// into the soul-planner backlog.

use regex::Regex;

use crate::db::{Result, TaskDB};
use crate::models::{Priority, TaskCreate};

// ---------------------------------------------------------------------------
// Block-to-priority mapping
// ---------------------------------------------------------------------------

/// Map block number to priority.
/// Block 1 (BUILD) = High, Block 2 (EXPLORE) = Normal,
/// Block 3 (SOCIAL) = Normal, Block 4 (SCOUT) = Low.
fn block_priority(block: u8) -> Priority {
    match block {
        1 => Priority::High,
        2 => Priority::Normal,
        3 => Priority::Normal,
        4 => Priority::Low,
        _ => Priority::Normal,
    }
}

// ---------------------------------------------------------------------------
// Month name mapping
// ---------------------------------------------------------------------------

fn month_number(name: &str) -> Option<u32> {
    match name.to_lowercase().as_str() {
        "jan" | "january" => Some(1),
        "feb" | "february" => Some(2),
        "mar" | "march" => Some(3),
        "apr" | "april" => Some(4),
        "may" => Some(5),
        "jun" | "june" => Some(6),
        "jul" | "july" => Some(7),
        "aug" | "august" => Some(8),
        "sep" | "september" => Some(9),
        "oct" | "october" => Some(10),
        "nov" | "november" => Some(11),
        "dec" | "december" => Some(12),
        _ => None,
    }
}

// ---------------------------------------------------------------------------
// Header parsers
// ---------------------------------------------------------------------------

/// Parse a date from a day header like "Day 2 -- Sun Feb 23".
///
/// Uses the current year since daily-planner.md does not include the year.
pub fn parse_date_from_header(header: &str) -> Option<chrono::NaiveDate> {
    let re = Regex::new(
        r"(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)\s+(\w+)\s+(\d+)"
    ).ok()?;
    let caps = re.captures(header)?;
    let month_str = &caps[1];
    let day_num: u32 = caps[2].parse().ok()?;
    let month_abbr = &month_str.to_lowercase()[..3.min(month_str.len())];
    let month_num = month_number(month_abbr)?;
    let year = chrono::Local::now().format("%Y").to_string().parse::<i32>().ok()?;
    chrono::NaiveDate::from_ymd_opt(year, month_num, day_num)
}

/// Extract block number from a header like "### Block 1: BUILD ...".
pub fn extract_block_number(header: &str) -> Option<u8> {
    let re = Regex::new(r"^###\s+Block\s+(\d+)").ok()?;
    let caps = re.captures(header)?;
    caps[1].parse().ok()
}

// ---------------------------------------------------------------------------
// PlannerTask
// ---------------------------------------------------------------------------

/// A parsed task from the daily planner.
#[derive(Debug, Clone)]
pub struct PlannerTask {
    pub title: String,
    pub block: u8,
    pub priority: Priority,
}

// ---------------------------------------------------------------------------
// parse_planner
// ---------------------------------------------------------------------------

/// Parse daily planner content and extract uncompleted tasks for a target date.
///
/// Behavior:
/// - Find `## Day N -- Weekday Month Date` section matching `target_date`
/// - Within that section, find `### Block N:` headers
/// - Extract `- [ ]` tasks (uncompleted only)
/// - Skip `[x]` (done), `[-]` (skipped), `[!]` (blocked)
/// - Stop at `---` (day separator)
/// - Apply `block_filter` if set
pub fn parse_planner(
    content: &str,
    target_date: chrono::NaiveDate,
    block_filter: Option<u8>,
) -> Vec<PlannerTask> {
    let task_re = Regex::new(r"^- \[ \] (.+)$").unwrap();
    let mut tasks = Vec::new();
    let mut in_target_day = false;
    let mut current_block: Option<u8> = None;

    for line in content.lines() {
        // Detect day headers: ## Day N -- Weekday Month Date
        if line.starts_with("## Day ") {
            let header_date = parse_date_from_header(line);
            in_target_day = header_date == Some(target_date);
            current_block = None;
            continue;
        }

        if !in_target_day {
            continue;
        }

        // Detect day separator (exit current day)
        if line.starts_with("---") {
            break;
        }

        // Detect block headers
        if line.starts_with("### Block ") {
            current_block = extract_block_number(line);
            continue;
        }

        // Non-block section headers (e.g., "### Evening Review") reset current_block
        if line.starts_with("### ") && !line.starts_with("### Block") {
            current_block = None;
            continue;
        }

        // Only process lines if we're inside a block
        let block = match current_block {
            Some(b) => b,
            None => continue,
        };

        // Apply block filter
        if let Some(filter) = block_filter {
            if block != filter {
                continue;
            }
        }

        // Extract uncompleted tasks: - [ ] task text
        let trimmed = line.trim();
        if let Some(caps) = task_re.captures(trimmed) {
            let title = caps[1].trim().to_string();
            tasks.push(PlannerTask {
                title,
                block,
                priority: block_priority(block),
            });
        }
    }

    tasks
}

// ---------------------------------------------------------------------------
// schedule_tasks_from_content
// ---------------------------------------------------------------------------

/// Parse planner content string and queue tasks to DB (skips duplicates).
///
/// This is the testable core of `schedule_tasks` -- accepts content as `&str`
/// rather than reading from a file.
pub fn schedule_tasks_from_content(
    db: &TaskDB,
    content: &str,
    target_date: chrono::NaiveDate,
    block_filter: Option<u8>,
) -> Result<Vec<i64>> {
    let parsed = parse_planner(content, target_date, block_filter);
    if parsed.is_empty() {
        return Ok(Vec::new());
    }

    // Build set of existing task titles for duplicate detection
    let existing = db.list(None)?;
    let mut existing_titles: std::collections::HashSet<String> =
        existing.into_iter().map(|t| t.title).collect();

    let mut created_ids = Vec::new();
    for item in &parsed {
        if existing_titles.contains(&item.title) {
            continue;
        }

        let create = TaskCreate {
            title: item.title.clone(),
            description: format!("From daily planner, Block {}", item.block),
            acceptance: String::new(),
            priority: item.priority,
            depends_on: Vec::new(),
        };
        let task = db.add(create)?;
        created_ids.push(task.id);
        existing_titles.insert(item.title.clone());
    }

    Ok(created_ids)
}

// ---------------------------------------------------------------------------
// schedule_tasks (file-based entry point)
// ---------------------------------------------------------------------------

/// Parse planner file and queue tasks to DB (skips duplicates).
///
/// Reads from `planner_path` (default `~/soul/docs/daily-planner.md`),
/// parses for `target_date` (default today), and creates tasks in the DB.
pub fn schedule_tasks(
    db: &TaskDB,
    planner_path: Option<&str>,
    target_date: Option<chrono::NaiveDate>,
    block_filter: Option<u8>,
) -> Result<Vec<i64>> {
    let default_path = dirs::home_dir()
        .map(|h| h.join("soul/docs/daily-planner.md"))
        .unwrap_or_else(|| std::path::PathBuf::from("docs/daily-planner.md"));
    let path = match planner_path {
        Some(p) => std::path::PathBuf::from(p),
        None => default_path,
    };

    if !path.exists() {
        return Err(format!("Planner file not found: {}", path.display()).into());
    }

    let content = std::fs::read_to_string(&path)?;
    let today = target_date.unwrap_or_else(|| chrono::Local::now().date_naive());

    schedule_tasks_from_content(db, &content, today, block_filter)
}
